### DISCUSSION TITLE
Store component properties for encoding

### DISCUSSION THREAD

**Discussion by Lloople** [Lloople/vapor-inertia-adapter#2] [Q&A]
I'll use an **edit user page** as an example for following the definitions along with my question.

## Definitions

- **Component**: Front-end component/page/template which uses the data (properties) to render the information, like `leaf` files but in `vue`, `react`, or whatever.
- **Property**: Data sent to the component, can be `String`, `Bool`, `Int`, `Model`, `Content`...
- **Inertia.swift**: This class stores `shared` properties like from a Middleware that would be sent to the component no matter which one it is. In the user edit page, it could be the logged-in user to show the name on the navbar, the app version to display on the footer...
- **InertiaResponse.swift**: This class prepares the response based on the headers to determine if it should return a leaf template or just JSON, it receives the name of the component (`User/Edit`) and the properties. In the user edit page, it would be the list of available languages in our app, the list of countries, a list of categories/topics to follow, etc...

I need to be able to store those properties both in Inertia and InertiaResponse and later on be merged and transformed to JSON

Both `Inertia` class and `InertiaResponse` class needs to store properties that would be transformed into JSON later on.

1. Middleware shares the user and the app version:

```swift
// This code goes in the project who uses this package. This is used to illustrate the use of Inertia shared properties
public struct InertiaSharedMiddleware: Middleware {
    public func respond(to request: Request, chainingTo next: Responder) -> EventLoopFuture<Response> {
        request.inertia.share("user", request.auth.has(User.self) ? request.auth.user : nil)
        request.inertia.share("appVersion", "v1.4.5")
    }
}
```

2. Controller receives the request and returns an InertiaResponse

```swift
// This code goes in the project who uses this package. This is used to illustrate the use of Inertia component properties
struct UserController {
    func edit(request: Request) throws -> EventLoopFuture<InertiaResponse> {
        return User.with(\.$company)
            .find(request.parameters.get("id", as: Int.self), on: request.db)
            .unwrap(or: Abort(.notFound))
            .map { user -> InertiaResponse in
                request.inertia.render("User/Edit", [
                    "languages": ["en", "es", "ca", "fr"],
                    "countries": ["uk", "es", "co", "ca", "us", "fr"]
                    "categories": ["News", "Videogames", "Synth", "Swift"]
                ])
    }
}
```

3. Inertia class [merges it's shared properties](https://github.com/Lloople/vapor-inertia-adapter/blob/discussion-1/Sources/VaporInertiaAdapter/Inertia.swift#L35) with the component ones before creating the InertiaResponse

4. When InertiaResponse is processed by Vapor,  method `encodeResponse` is called transforming the properties into JSON and returning a [JSON response](https://github.com/Lloople/vapor-inertia-adapter/blob/discussion-1/Sources/VaporInertiaAdapter/InertiaResponse.swift#L44) or a [Leaf view template](https://github.com/Lloople/vapor-inertia-adapter/blob/discussion-1/Sources/VaporInertiaAdapter/InertiaResponse.swift#L34) depending on request headers.

This is an example of JSON returned 

```json
{
  "component": "Dashboard/Index",
  "props": {
    "errors": {},
    "auth": {
      "user": {
        "id": 1,
        "first_name": "John",
        "last_name": "Doe",
        "email": "johndoe@example.com",
        "role": null,
        "account": {
          "id": 1,
          "name": "Acme Corporation"
        }
      }
    },
    "flash": {
      "success": null,
      "error": null
    }
  },
  "url": "/",
  "version": "4d0624a81db51088551dfc347a878621"
}
```

**Comment by Steven4294** [comment]
Why not just store that info client side if its not changing between requests?

**Reply by Lloople** [reply]
Because I’m building an adapter for an existent JavaScript library 😅 

Also the info could change between requests. It’s like a server side app but the refresh happens in background via Ajax to feel like a SPA
